lib/active_merchant/billing/gateways/worldpay.rb

```diff
@@ -483,9 +548,52 @@ def required_status_message(raw, success_criteria)
         end
       end
 
-      def authorization_from(raw)
-        pair = raw.detect { |k, v| k.to_s =~ /_order_code$/ }
-        (pair ? pair.last : nil)
+      def authorization_from(action, raw, options)
+        case action
+        when 'store'
+          authorization_from_token_details(
+            token_id: raw[:payment_token_id],
+            token_scope: 'shopper',
+            customer: options[:customer]
+          )
+        else
+          pair = raw.detect { |k, v| k.to_s =~ /_order_code$/ }
+          (pair ? pair.last : nil)
+        end
+      end
+
```

Comment:
If, in the future, anyone needs to support merchant-scoped tokens, it will be easy to add without breaking compatibility with the current format. Likewise should this form of authorization need to embed the original order id (not something we need to do now, but would be needed to be able to return an authorization from a authorize/purchase call where the caller also wants to store the token in a single call (Worldpay does support this use case)).

Reply:
So I think I like this. But I do want to bring up a consideration. Obviously the way AM handles `authorization`s is not ideal, but to the extent that there is a standard, that standard seems to be for newly supported elements to be concatenated on the end of whatever is already present, for any action type. So what I would have expected here is that the data for a stored card would start with the second element of a concatenated authorization, regardless of whether that first element actually has a value.

Does that make sense? And if so, does it sound like something we should actually care about?

Reply:
The advantage of this is that, if we're assuming it needs to remain a string, the ordering allows it to be parsed both by the adapter and by anything else looking at the string (as long as the order is known and maintained), independent of transaction type (some authorizations add the action type for future explicit parsing).

Reply:
Yah, I think it's a good idea (and easy to implement) to ensure that the order_id, if ever it should be present in a compound auth like this one, would take the first ordinal position (and just be empty for now in store where it doesn't apply). That way any future extensions where both the order_id and these token-specific bits are necessary (such as if anyone needs to implement authorize+store in a single call) can rely on the first element (including if there are no "|" characters) always being the order_id.

Not completely sure though if I follow about checking the length of split for pay-as-order though. Are you referring to how we distinguish between pay-as-order payment method strings and stored token id strings?

Reply:
Cool, I think we should go with that.

>Not completely sure though if I follow about checking the length of split for pay-as-order though. Are you referring to how we distinguish between pay-as-order payment method strings and stored token id strings?

Yes, though I guess it depends on how the current methods change, we might not even need that. I'll leave it up to you and wait to see the changes.

Reply:
@curiousepic Ok, I've pushed up some updates that implement this.

Since we want to ensure there is a slot in our compound authorization strings that can hold the `order_id` value (if and when it's used), I went ahead and ensured that these complex authorization strings could be used in place of a bare order_id authorization.

Was actually a bit uncertain whether to proceed that far, but it would have felt awkward to me to have the ability to store the order_ids in an authorization that would have been invalid to use for capture/etc. Especially welcome feedback on this point.